main/lsp/ExtractVariable.cc

```diff
@@ -268,4 +289,216 @@ vector<unique_ptr<TextDocumentEdit>> VariableExtractor::getEdits(LSPTypecheckerD
+        }
+    }
+    void preTransformInsSeq(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformInsSeq(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
+
+    void preTransformClassDef(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformClassDef(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
+
+    void preTransformMethodDef(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformMethodDef(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
+
+    void preTransformBlock(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformBlock(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
+
+    void preTransformIf(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformIf(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
+
+    void preTransformRescue(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformRescue(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
+
+    void preTransformWhile(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformWhile(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
+};
+
+std::pair<vector<unique_ptr<TextDocumentEdit>>, int> VariableExtractor::getExtractMultipleOccurrenceEdits() {
+    ENFORCE(matchingNode, "getExtractMultipleOccurrenceEdits called before getExtractSingleOccurrenceEdits");
+    ENFORCE(enclosingClassOrMethod, "getExtractMultipleOccurrenceEdits called before getExtractSingleOccurrenceEdits");
+
+    const auto file = selectionLoc.file();
+    const auto &gs = typechecker.state();
+
+    ExpressionPtrSearchWalk walk(&matchingNode, skippedLocs);
+    core::Context ctx(gs, core::Symbols::root(), file);
+    ast::TreeWalk::apply(ctx, walk, enclosingClassOrMethod);
+
+    auto matches = walk.matches;
+
+    // There should be at least one match (the original selected expression).
+    ENFORCE(matches.size() > 0);
+    if (matches.size() == 1) {
+        return std::pair(vector<unique_ptr<TextDocumentEdit>>(), 1);
+    }
+    fast_sort(
+        matches, [](auto a, auto b) -> auto { return a.second->loc().beginPos() < b.second->loc().beginPos(); });
+
+    // Find LCA enclosing scope
```

Comment:
Could this algorithm be substantially simplified if it were agnostic of node type, and instead used pointer equality?

For example, given

```
matches = [
  [0x1, 0x2, 0x3],
  [0x1, 0x2, 0x4],
  [0x1, 0x5, 0x6],
]
```

We could loop over the array (in column major order, which is a bit weird because you'd need to know the max sub-vector's length to do that), and say "are all of these enclosing scopes the same (by pointer equality) at this depth? if so, this is our new LCA, go to the next index. Otherwise, break out of the loop."

And at that point, is there any reason why we couldn't move this computation to be done **during** the tree walk? That way, we would only need to keep "current tightest LCA stack" and "current enclosing scope stack", which uses memory linear in the size of the tree, not quadratic in the size of the tree.

Reply:
Yeah I think this makes sense. Thanks!